INT-4051: Add PublisherAnnBPP when `@EnableInt`

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4051

Previously we can process `@Publisher` annotations only with the `<int:annotation-config>` or with an explicit `@EnablePublisher("")`
- Fix `PublisherRegistrar` to relax from the fact if we have `@EnablePublisher` or not and register `PublisherAnnotationBeanPostProcessor` without `defaultChannelName`

artembilan committed:
INT-4051: Add PublisherAnnBPP when `@EnableInt`

JIRA: https://jira.spring.io/browse/INT-4051

Previously we can process `@Publisher` annotations only with the `<int:annotation-config>` or with an explicit `@EnablePublisher("")`

* Fix `PublisherRegistrar` to relax from the fact if we have `@EnablePublisher` or not and register `PublisherAnnotationBeanPostProcessor` without `defaultChannelName`

artembilan commented:
I think you can just let me know if you are fine with the solution and I'll merge myself.
No reason to torment yourself with push permissions if that :smile:

mbogoevici commented on spring-integration-core/src/test/java/org/springframework/integration/config/annotation/MessagingAnnotationsWithBeanAnnotationTests.java:
```diff
@@ -145,6 +150,8 @@ public void testMessagingAnnotationsFlow() {
 		assertNull(this.skippedChannel2);
 		assertNull(this.skippedMessageSource);
+
+		assertNotNull(this.publisherChannel.receive(10000));
```
Maybe it would help to check that the message received on the publisher channel is the one we expect? I know that this would be in a way re-testing the `@Publisher` annotation, I think of it more as defense against someone inadvertently altering the tests in the future - e.g. adding a second `@Publisher` test and so on.

mbogoevici commented on spring-integration-core/src/main/java/org/springframework/integration/config/PublisherRegistrar.java:
```diff
@@ -46,10 +46,8 @@
-		}
-		String value = (String) annotationAttributes.get("value");
+
+		String value = (annotationAttributes == null ? null : (String) annotationAttributes.get("value"));
```
Perhaps it would be better to do `AnnotationUtils.getDefaultValue(EnablePublisher.class)` - this would protect against annotation changes in the future.

mbogoevici commented:
LGTM with minor comments.

One thought for the future would be perhaps to expand `@EnablePublisher` with a `proxyTargetClass` parameter, that could enable class/CGLib proxies. With classes without interfaces become more common with JavaConfig, this could be a useful simplification.

artembilan committed:
Address PR Comments

artembilan commented:
Pushed.
I have changed my mind and moved test to the `BarrierMessageHandlerTests`. Because `MessagingAnnotationsWithBeanAnnotationTests` is enough complex to track all the received messages.

Feel free to raise `proxyTargetClass` JIRA for 5.0. There are need some tricks to make it clear for XML config. No problem with `@EnablePublisher` though!

mbogoevici commented:
LGTM. Merging.

mbogoevici commented:
Squashed and merge w/https://github.com/spring-projects/spring-integration/commit/cabe507e0ef7cd2be830e2a07ef45da057e1fb33
